### PR TITLE
113

### PR DESCRIPTION
**Pull request by rombautt**

### CODE REVIEW

File: foundations/src/main/java/io/yields/math/framework/definition/AlgorithmDefinition.java [hunk 1]
```diff
@@ -33,6 +33,12 @@
     private Map<String, String> attributes;
     private Class<? extends ExplorationContext<T, V>> explorationContext;
 
+    /**
+     * For JSON
+     */
+    public AlgorithmDefinition() {
```
Comment (jgheerardyn) [inline foundations/src/main/java/io/yields/math/framework/definition/AlgorithmDefinition.java]: Can we make these empty constructors *protected* or *private*?

File: foundations/src/main/java/io/yields/math/framework/definition/junit/JUnitResultFormatterRunListener.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+    @Override
+    public void testAssumptionFailure(Failure failure) {
+        problem++;
+        TestDescription testDescription = new TestDescription(substringBefore(failure.getDescription().getDisplayName(), "("));
+        formatter.addError(new JUnitReportEntry(testDescription.getTest()), failure.getException());
+    }
+
+    @Override
+    public void testIgnored(Description description) throws Exception {
```
Comment (jgheerardyn) [inline foundations/src/main/java/io/yields/math/framework/definition/junit/JUnitResultFormatterRunListener.java]: is there a method that is called when a test throws an (un)expected exception?

File: foundations/src/main/java/io/yields/math/framework/definition/junit/TestDefinitionJUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        logger.info("Started exploring " + explorer);
+
+        long start = System.currentTimeMillis();
+        long totalResults = explorer.getExplorerBatchBatches().stream().mapToLong(explorerBatch -> explorerBatch.all().count()).sum();
+        long end = System.currentTimeMillis();
+
+        logger.info(String.format("Completed exploring %s in %dms, with %d results", explorer, end - start, totalResults));
+    }
+
+    public static class TestExplorationContext implements ExplorationContext<Double, Double> {
```
Comment (jgheerardyn) [inline foundations/src/main/java/io/yields/math/framework/definition/junit/TestDefinitionJUnitTest.java]: how come this is <Double,Double> and not generically parameterized?

Comment (jgheerardyn) [inline foundations/src/main/java/io/yields/math/framework/definition/junit/TestDefinitionJUnitTest.java]: how come this is <DoubleDouble> and not generically parameterized?

**Comment by jgheerardyn** [comment]
Can you add some explanation on the Wiki on how to use this?

File: foundations/src/main/java/io/yields/math/framework/definition/AlgorithmDefinition.java (hunk 1 above)
Comment (rombautt) [inline foundations/src/main/java/io/yields/math/framework/definition/AlgorithmDefinition.java]: yes, made them private.

File: foundations/src/main/java/io/yields/math/framework/definition/junit/JUnitResultFormatterRunListener.java (hunk 2 above)
Comment (rombautt) [inline foundations/src/main/java/io/yields/math/framework/definition/junit/JUnitResultFormatterRunListener.java]: I would expect this is also captured with testFailure: the JUnit model works with failure by throwing an AssertionError, this is no different than another Exception.

File: foundations/src/main/java/io/yields/math/framework/definition/junit/TestDefinitionRunner.java [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+            }
+        });
+        return description;
+    }
+
+    public static void main(String... args) throws Exception {
+
+        Options options = new Options();
+        options.addOption("config", true, "JSON with test definitions to execute");
+        options.addOption("configFile", true, "JSON file with test definitions to execute");
```
Comment (jgheerardyn) [inline foundations/src/main/java/io/yields/math/framework/definition/junit/TestDefinitionRunner.java]: so two options with to support the same functionality?

File: foundations/src/main/java/io/yields/math/framework/definition/junit/TestDefinitionRunner.java [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+        System.setProperty(TEST_DEFINITIONS, testDefinitions);
+
+        File reportDir = new File(cmd.getOptionValue("output"));
+        deleteDirectory(reportDir);
+        forceMkdir(reportDir);
+
+        junit.addListener(new JUnitResultFormatterRunListener(new XMLJUnitResultFormatter(), reportDir));
+
+        Result run = junit.run(new Computer(), TestDefinitionJUnitTest.class);
```
Comment (jgheerardyn) [inline foundations/src/main/java/io/yields/math/framework/definition/junit/TestDefinitionRunner.java]: TestDefinition.TestConfiguration contains a minimal score value. I can't find the place where this minimal score is asserted inside the junit test. Did I miss that code?

Comment (rombautt) [inline foundations/src/main/java/io/yields/math/framework/definition/junit/TestDefinitionRunner.java]: K, did not know about this, I'll add assers

File: foundations/src/main/java/io/yields/math/framework/definition/junit/TestDefinitionRunner.java (hunk 4 above)
Comment (rombautt) [inline foundations/src/main/java/io/yields/math/framework/definition/junit/TestDefinitionRunner.java]: That's true: you can run with data from a file, however, testpad will kick builds with the data in a string, so both are supported.

**Pull request closed without merging**